### PR TITLE
[593] Migrate and rework the entire vacancies editing functionality

### PR DESCRIPTION
**Pull request by darokel**
### Context

- https://trello.com/c/OC8lv6IK/593-migrate-courses-edit-vacancies

### Changes proposed in this pull request

- Refactor and simplify entire editing vacancies flow
- Use govuk formbuilder for views over manual markup
- Move logic over to form object pattern
- Increase test coverage with feature spec

### Guidance to review

Edit vacancies for a course with one site https://teacher-training-api-pr-2490.london.cloudapps.digital/publish/organisations/2AT/2022/courses

Edit vacancies for a course with a site and multiple study modes https://teacher-training-api-pr-2490.london.cloudapps.digital/publish/organisations/2EZ/2022/courses (33PY)

Edit vacancies for a course with multiple sites https://teacher-training-api-pr-2490.london.cloudapps.digital/publish/organisations/2CJ/2022/courses (M444)

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

**Comment by darokel** [comment]
[Migrate Courses - Edit Vacancies](https://trello.com/c/OC8lv6IK/593-migrate-courses-edit-vacancies)

**Comment by darokel** [comment]
A bit of a whopper - I was going down a rabbit hole of potentially doing too much refactoring at the cost of time. There are a still a few things to clean up like putting the enum values into re-usable constants amongst other things. We can do all of this after the migration is done.

File: config/rubocop/naming.yml [hunk 1]
```diff
@@ -1,2 +1,4 @@
 Naming/MemoizedInstanceVariableName:
   EnforcedStyleForLeadingUnderscores: optional
+  Exclude:
+  - "spec/**/*"
```
[the remark this answers is not in the document]
Comment (darokel) [reply config/rubocop/naming.yml:4]: I think we need this to give us flexibility in writing re-usable steps for fixtures. We had it disabled in Register too

File: app/controllers/publish_interface/providers/courses/vacancies_controller.rb [hunk 2]
```diff
@@ -1,9 +1,62 @@
 module PublishInterface
   module Providers
-    class VacanciesController < PublishInterfaceController
-      def edit; end
+    module Courses
+      class VacanciesController < PublishInterfaceController
+        def edit
+          authorize(provider, :edit?)
```
[the remark this answers is not in the document]
Comment (darokel) [reply app/controllers/publish_interface/providers/courses/vacancies_controller.rb]: done https://github.com/DFE-Digital/teacher-training-api/pull/2490/commits/34616e7b9ee3b872c86b9698a87e468ac5eddd87

File: app/controllers/publish_interface/providers/courses/vacancies_controller.rb [hunk 3]
```diff
@@ -1,9 +1,62 @@
...
+        end
+
+      private
+
+        def course
+          @course ||= provider.courses.find_by!(course_code: params[:id])
+        end
+
+        def provider
+          @provider ||= Provider.find_by!(recruitment_cycle: recruitment_cycle, provider_code: params[:provider_code])
```
[the remark this answers is not in the document]
Comment (darokel) [reply app/controllers/publish_interface/providers/courses/vacancies_controller.rb]: i've streamlined https://github.com/DFE-Digital/teacher-training-api/pull/2490/commits/34616e7b9ee3b872c86b9698a87e468ac5eddd87

File: app/controllers/publish_interface/providers/courses/vacancies_controller.rb [hunk 4]
```diff
@@ -1,9 +1,62 @@
...
+        end
+
+        def recruitment_cycle
+          cycle_year = params[:recruitment_cycle_year] || params[:year] || Settings.current_recruitment_cycle_year
```
[the remark this answers is not in the document]
Comment (darokel) [reply app/controllers/publish_interface/providers/courses/vacancies_controller.rb]: yeah the middle. i think we should just define these once in the publish interface controller thingy - will make a ticket for it

File: spec/support/page_objects/publish_interface/course_vacancies_edit.rb [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+        within("#publish-interface-course-vacancies-form-has-vacancies-true-conditional") do
+          all(".govuk-checkboxes__item")
+        end
+      end
+
+      def site_checkboxes
+        within("#publish-interface-course-vacancies-form-has-vacancies-true-conditional") do
+          all(".govuk-checkboxes__input")
+        end
+      end
```
[the remark this answers is not in the document]
Comment (darokel) [reply spec/support/page_objects/publish_interface/course_vacancies_edit.rb]: cleaned this up https://github.com/DFE-Digital/teacher-training-api/pull/2490/commits/34616e7b9ee3b872c86b9698a87e468ac5eddd87

File: spec/support/page_objects/publish_interface/course_vacancies_edit.rb [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+      element :vacancies_radio_no_vacancies, "#publish-interface-course-vacancies-form-has-vacancies-field"
+      element :vacancies_radio_has_some_vacancies, "#publish-interface-course-vacancies-form-has-vacancies-true-field"
+
+      element :submit, 'button.govuk-button[type="submit"]'
+
+      def errors
+        within(error_summary) do
+          all(".govuk-error-summary__list li").map(&:text)
+        end
+      end
```
[the remark this answers is not in the document]
Comment (darokel) [reply spec/support/page_objects/publish_interface/course_vacancies_edit.rb]: done https://github.com/DFE-Digital/teacher-training-api/pull/2490/commits/34616e7b9ee3b872c86b9698a87e468ac5eddd87

File: app/controllers/publish_interface/providers/courses/vacancies_controller.rb [hunk 7]
```diff
@@ -1,9 +1,62 @@
...
+
+        def vacancy_params
+          return { change_vacancies_confirmation: nil } if params[:publish_interface_course_vacancies_form].blank?
+
+          params
+            .require(:publish_interface_course_vacancies_form)
+            .permit(
+              CourseVacanciesForm::FIELDS,
+              :change_vacancies_confirmation,
+              :has_vacancies,
```
[the remark this answers is not in the document]
Comment (darokel) [reply app/controllers/publish_interface/providers/courses/vacancies_controller.rb]: fixed https://github.com/DFE-Digital/teacher-training-api/pull/2490/commits/34616e7b9ee3b872c86b9698a87e468ac5eddd87

File: app/views/publish_interface/providers/courses/vacancies/_multiple_sites.html.erb [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+  <%= f.govuk_radio_button(:has_vacancies, true, label: { text: "There are some vacancies", size: "s" },
+    hint: { text: "Select the locations with vacancies" },
+    link_errors: true) do %>
+
+    <% @site_statuses.sort_by { |status| status.site.location_name }.each do |site_status| %>
+      <%= f.fields_for :site_statuses, site_status do |site_status| %>
+        <% if @course.full_time_or_part_time? %>
+          <%= render partial: "site_status_checkbox", locals: { f: site_status, course: @course, study_mode: :part_time, show_study_mode: true } %>
+          <%= render partial: "site_status_checkbox", locals: { f: site_status, course: @course, study_mode: :full_time, show_study_mode: true } %>
+        <% elsif @course.full_time? || @course.part_time? %>
```
Comment (defong) [inline app/views/publish_interface/providers/courses/vacancies/_multiple_sites.html.erb:18]: is this `if` actually needed?

File: app/forms/publish_interface/course_vacancies_form.rb [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+        .merge(has_vacancies: course.has_vacancies?)
+    end
+
+    def assign_attributes_to_model
+      model.assign_attributes(
+        fields
+          .symbolize_keys
+          .except(*fields_to_ignore_before_save)
+          .deep_merge(attributes_for_site_statuses),
+      )
```
Comment (defong) [inline app/forms/publish_interface/course_vacancies_form.rb:43]: It looks a little be wieldy, can we change it to something like instead please
```ruby
      model.assign_attributes(model_attributes)

...

      def model_attributes
        fields
          .symbolize_keys
          .except(*fields_to_ignore_before_save)
          .deep_merge(attributes_for_site_statuses)
      end
```

File: spec/factories/courses.rb [hunk 10]
```diff
@@ -198,6 +198,14 @@
       to_create { |instance| instance.save(validate: false) }
     end
 
+    trait :full_time_or_part_time do
+      study_mode { :full_time_or_part_time }
+    end
+
+    trait :part_time do
+      study_mode { "part_time" }
+    end
```
Comment (defong) [inline spec/factories/courses.rb:207]: Unnecessary? 
https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#enum-traits

File: app/models/course.rb [hunk 11]
```diff
@@ -393,6 +394,14 @@ def has_vacancies?
     end
   end
 
+  def has_multiple_running_sites_or_study_modes?
+    running_site_statuses.length > 1 || full_time_or_part_time?
+  end
+
+  def running_site_statuses
+    site_statuses.select(&:running?)
```
Comment (defong) [inline app/models/course.rb:402]: Has this made it really expensive as it outside or active record land?

File: spec/support/page_objects/sections/vacancy.rb [hunk 12]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+require_relative "base"
+require "forwardable"
+
+module PageObjects
+  module Sections
+    class Vacancy < PageObjects::Sections::Base
```
Comment (defong) [inline spec/support/page_objects/sections/vacancy.rb:8]: Maybe we should just call this GovCheckbox, what do you think?

File: app/models/course.rb (hunk 11 above)
Comment (darokel) [reply app/models/course.rb:402]: This was copy pasta'd from old publish code and I was meant to refactor after adding coverage - good spot

File: app/views/publish_interface/providers/courses/vacancies/_multiple_sites.html.erb (hunk 8 above)
Comment (darokel) [reply app/views/publish_interface/providers/courses/vacancies/_multiple_sites.html.erb]: remooved!

File: app/forms/publish_interface/course_vacancies_form.rb (hunk 9 above)
Comment (darokel) [reply app/forms/publish_interface/course_vacancies_form.rb]: agreed, updated

File: app/models/course.rb (hunk 11 above)
Comment (darokel) [reply app/models/course.rb]: fixed

File: spec/support/page_objects/sections/vacancy.rb (hunk 12 above)
Comment (darokel) [reply spec/support/page_objects/sections/vacancy.rb:8]: will do so when we need to re-use it in another page

File: spec/factories/courses.rb (hunk 10 above)
Comment (darokel) [reply spec/factories/courses.rb]: yup, removed

**Review by defong — APPROVED** [review]
Awesome work

**Review by mbocc — APPROVED** [review]
All works, apologies for tardiness in reviewing this. Fab work, thank you!

**Pull request merged**